### ISSUE TITLE
Optimiser changes order of the gates

### ISSUE THREAD

**Issue by kevaundray** [noir-lang/noir#109]
The following circuit will ICE the compiler with a witness assignment not found error. This only happens if the gates are out of order, since we have the compiler guaranteeing that variables are both initialised properly and used.

```rust
use dep::std;

fn main(x : Field) {
    
    priv y = x + x + x;

   std::hash::pedersen([y]);
}
```

Inspecting the unoptimised ACIR, we see this:

```rust
x = Witness(1)
y = Witness(2)

[
        Arithmetic {
            mul_terms: [],
            linear_combinations: [
                (
                    FieldElement(1),
                    Witness(
                        1,
                    ),
                ),
                (
                    FieldElement(1),
                    Witness(
                        1,
                    ),
                ),
                (
                    FieldElement(1),
                    Witness(
                        1,
                    ),
                ),
                (
                    FieldElement(-1),
                    Witness(
                        2,
                    ),
                ),
            ],
            q_c: FieldElement(0),
        },
        GadgetCall {
            name: Pedersen,
            inputs: [
                GadgetInput {
                    witness: Witness(
                        2,
                    ),
                    num_bits: 254,
                },
            ],
            outputs: [
                Witness(
                    3,
                ),
            ],
        },
    
]
```

This maps to the above program as expected, which prompted the suspicion that the optimiser was disordering the gates. This can be seen here: https://github.com/noir-lang/noir/blob/master/crates/acvm/src/compiler.rs#L40

When the optimiser optimises a gate, it is added to a separate list of gates, and then re-added to the back of the list of gates. Instead the optimiser should insert the gate inplace.

**Comment by kevaundray** [comment]
Digging deeper.

Consider the following code:

```rust
priv y = x + x + x
```

For width 3 , we should get the following gates:

```
0 = x + x + x + y;
0 = x + x - tmp; (Intermediate gate)
0 = tmp + x + y
```

Note that the following will pass the witness generator.

What we actually get is:

```
0 = x + x + x + y;
0 = x + y - tmp; (Intermediate gate)
0 = x + x + tmp
```

The second equation is not solveable, without looking at the third, as it has two unknowns. 

## Solution 1

We can solve this by implementing a way for the witness generator to solve systems of equations, instead of one equation at a time.

## Solution 2

We can re-order the variables in the gate, so that the unknown variable is at the start of the vector and so they are used last when creating temporary variables.

I will open a new issue for solution 1, and opt for solution 2 right now. The only place where we need this is for private variable declaration, so solution 2 is small in scope.

The problem in particular is here: https://github.com/noir-lang/noir/blob/master/crates/acir/src/optimiser/csat_optimiser.rs#L292

Where we are creating a new intermediate variable to collapse the width, but we can pop off an unknown

**Comment by kevaundray** [comment]
closed by #110
